Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp

```diff
@@ -0,0 +1,475 @@
+{
+    m_hover_over_control_indicator = false;
+    update();
+}
+
+void DynamicWidgetContainer::resize_event(ResizeEvent&)
+{
+    constexpr int button_width_and_internal_margin = 24;
+    m_expand_collapse_button.set_x(width() - (button_width_and_internal_margin + content_margins().right()));
+    m_detach_widget_button.set_x(m_expand_collapse_button.x() - 18);
+
+    // Check if there is any content to display, and hide ourself if there would be nothing to display.
+    // This allows us to make the whole section not taking up space if child-widget visibility is maintained outside.
+    if (m_previous_frame_style.has_value() && height() != 0) {
+        m_controls_spacer->set_visible(true);
+        set_frame_style(m_previous_frame_style.value());
+        m_previous_frame_style = {};
+
+        deferred_invoke([&]() {
+            invalidate_layout();
+        });
+    }
+
+    if (view_state() == ViewState::Expanded && !m_previous_frame_style.has_value() && height() == (content_margins().top() + content_margins().bottom() + m_controls_spacer->height())) {
+        m_controls_spacer->set_visible(false);
+        m_previous_frame_style = frame_style();
+        set_frame_style(Gfx::FrameStyle::NoFrame);
+
+        deferred_invoke([&]() {
+            invalidate_layout();
+        });
+    }
+}
+
+void DynamicWidgetContainer::child_event(Core::ChildEvent& event)
+{
+    if (event.type() == Event::ChildAdded && event.child() && is<GUI::DynamicWidgetContainer>(*event.child()))
+        static_cast<DynamicWidgetContainer&>(*event.child()).set_parent_container(*this);
+
+    GUI::Frame::child_event(event);
+}
+
+void DynamicWidgetContainer::set_parent_container(RefPtr<GUI::DynamicWidgetContainer> container)
+{
+    m_parent_container = container;
+}
+
+bool DynamicWidgetContainer::check_has_move_target(Gfx::IntPoint relative_mouse_position, MoveTargetOp operation)
+{
+    bool matched = false;
+    for_each_child_container([&](auto& child) {
+        bool is_target = child.relative_rect().contains(relative_mouse_position);
+        matched |= is_target;
+        child.set_render_as_move_target(operation == MoveTargetOp::SetTarget ? is_target : false);
+
+        return IterationDecision::Continue;
+    });
+    return matched;
+}
+
+void DynamicWidgetContainer::set_render_as_move_target(bool is_target)
+{
+    if (m_render_as_move_target == is_target)
+        return;
+    m_render_as_move_target = is_target;
+    update();
+}
+
+void DynamicWidgetContainer::swap_widget_positions(RefPtr<Core::EventReceiver> source_widget, Gfx::IntPoint destination_positon)
+{
+    RefPtr<Core::EventReceiver> destination_widget;
+    for_each_child_container([&](auto& child) {
+        if (child.relative_rect().contains(destination_positon)) {
+            destination_widget = child;
+            return IterationDecision::Break;
+        }
+        return IterationDecision::Continue;
+    });
+
+    VERIFY(destination_widget);
+    VERIFY(source_widget);
```

Comment:
Instead pass the source widget as a NNRP; the caller must check that it's passing a valid widget.